utils/show.cc

```diff
@@ -399,11 +405,136 @@ void doScrape(tr_torrent_metainfo const& metainfo)
     }
 }
 
+namespace verify_cb
+{
+
+struct Status
+{
+    tr_file_index_t index;
+    bool ok;
+    std::vector<std::string> errors;
+};
+
+std::vector<Status> files;
+tr_file_index_t current_index;
+std::vector<std::string> errors;
+bool print_ok_files;
+bool unsorted;
+tr_torrent_metainfo const* metainfo;
+
+void print_file(Status const& status)
+{
+    fmt::print(
+        "  {:s} {:s} {:s}\n",
+        status.ok ? "OK  " : "FAIL",
+        metainfo->file_subpath(status.index),
+        tr_formatter_size_B(metainfo->file_size(status.index)));
+    for (auto const& file_error : status.errors)
+    {
+        fmt::print("    {}\n", file_error);
+    }
+}
+
+void verify_callback(tr_file_index_t index, bool ok, std::string error)
+{
+    if (index != current_index)
```

Comment:
Please correct me if I am wrong here.

AFAICT, `verify_callback()` is called at most once for each file, so this if condition isn't needed. In fact, this is causing the last file to *not* be printed, which you worked around by repeating code inside `verify()`, so there's a very good reason to remove it.

Following the same logic, we can also do away with `verify_cb::errors` because each file will have at most 1 error getting printed out. `verify_callback()` can probably be drastically simplified after this change.

Reply:
When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. I could make it one callback per file with a vector of errors but it makes the `verify_torrent()` logic more complicated. It doesn't help that a piece can contain multiple files.

Reply:
> When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. 

Does it? I thought the error callback will only be called for the first piece that errored, because `file_valid` will be set to false here:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L140

And so the file won't be added to `files_ending_in_current_piece` for the pieces afterwards.

> I could make it one callback per file with a vector of errors but it makes the `verify_torrent()` logic more complicated. It doesn't help that a piece can contain multiple files.

No no, that's not what I want either. In that case, how about you initialise `current_index` to `~0` instead of `0`, and set `status.index` to `index` instead of `current_index` inside `verify_callback()`?

```cpp
// Before
tr_file_index_t current_index;
// After
tr_file_index_t current_index = ~0;

// Before
status.index = current_index;
// After
status.index = index;
```

Reply:
> When there are errors `verify_callback()` can be called multiple times per file. eg Multiple failed pieces in the one file. 

Oh! I see it now. Yes you are right.

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L136-L139

Reply:
I was thinking that if any piece fails we could just report "hash fail" for the affected file(s) instead of reporting each piece that failed. This would simplify the code as there would only be one callback per file - either success, some type of file error (unable to open, wrong size, etc), or hash fail. I could merge `on_file_ok` and `on_file_error` into one callback.

I don't think it makes much difference to the end user whether we report only that a file failed the hash check or report every piece that failed. Either way the user can't trust the file.

Reply:
I think either way is acceptable, feel free to write whatever that's comfortable for you.

I'd say keep the current behaviour though, since you've already put in the effort, and the fix for the repeated code seems to be very simple:

> In that case, how about you initialise `current_index` to `~0` instead of `0`, and set `status.index` to `index` instead of `current_index` inside `verify_callback()`?